harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+
+
+    def _begin_collection(self):
+        self.sys_metric_collector_thread.activate()
+        # TODO: Start up TimingBatcher as well
+
+    def _end_collection(self):
+        self.sys_metric_collector_thread.kill()
+        self.sender_thread.kill()
+        # TODO: Shut down TimingBatcher as well
+
+        self.sys_metric_collector_thread.join()
+        self.sender_thread.join()
+
+
+    def record_timing(self):
+        if not self.is_enabled:
+            return
+        # TODO: Add new timing to TimingBatcher
+
+
+
+class Measurement:
+    def __init__(self, timestamp: datetime.datetime, batch_idx, value):
+        self.timestamp = timestamp
+        self.batch_idx = batch_idx
+        self.measurement = value
+
+
+class SysMetricType:
+    GPU_UTIL_METRIC = "gpu_util"
+    NET_THRU_SENT_METRIC = "net_throughput_sent"
+    NET_THRU_RECV_METRIC = "net_throughput_recv"
+    DISK_IOPS_METRIC = "disk_iops"
+    DISK_THRU_READ_METRIC = "disk_throughput_read"
+    DISK_THRU_WRITE_METRIC = "disk_throughput_write"
+    FREE_MEM_METRIC = "free_memory"
+    SIMPLE_CPU_UTIL_METRIC = "cpu_util_simple"
+
+
+class SysMetricBatcher:
+    """
+    Data structure to collect batches of SysMetrics and then convert them to the format expected by the API
+    """
+
+    def __init__(self, trial_id, agent_id):
+        self.trial_id = trial_id
+        self.agent_id = agent_id
+        self.clear()
+
+    def clear(self):
+        self.batch = {
+            SysMetricType.GPU_UTIL_METRIC: {},
+            # "GPU_MEM": [],
+            SysMetricType.NET_THRU_SENT_METRIC: [],
+            SysMetricType.NET_THRU_RECV_METRIC: [],
+            # "DISK_FREE": [],
+            SysMetricType.DISK_IOPS_METRIC: [],
+            SysMetricType.DISK_THRU_READ_METRIC: [],
+            SysMetricType.DISK_THRU_WRITE_METRIC: [],
+            SysMetricType.FREE_MEM_METRIC: [],
+            SysMetricType.SIMPLE_CPU_UTIL_METRIC: []
+        }
+
+    def add_nongpu_measurement(self, metric_type, measurement):
+        assert metric_type in self.batch.keys(), \
+            f"Tried to add unknown type of non-GPU metric: {metric_type}"
+        self.batch[metric_type].append(measurement)
+
+    def add_gpu_measurement(self, metric_type, gpu_uuid, measurement):
+        assert metric_type in self.batch.keys(), \
+            f"Tried to add unknown type of GPU metric: {metric_type}"
+        if gpu_uuid not in self.batch[metric_type].keys():
+            self.batch[metric_type][gpu_uuid] = []
+        self.batch[metric_type][gpu_uuid].append(measurement)
+
+    def convert_to_timestamp_str(self, timestamp: datetime.datetime):
+        assert isinstance(timestamp, datetime.datetime), \
+            f"Input to conversion function must be a datetime object. Instead got {type(timestamp)}"
+        return timestamp.isoformat() + "Z"
+
+    def convert_to_post_format(self):
+        post_formatted = []
+        for metric_type in self.batch.keys():
```

Comment:
This seems a little involved, I think it's a symptom of the `add_nongpu_measurement`/`add_gpu_measurement` dichotomy. It may make more sense/save a lot of effort to have the `Measurement` type have a `gpu_uuid` that is filled in some cases and not in others and then just let the rest of the code behave the same.